hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/AbstractContainerReportHandler.java

```diff
@@ -352,26 +352,32 @@ private boolean updateContainerState(final DatanodeDetails datanode,
        */
       break;
     case DELETING:
+    case DELETED:
       /*
-      HDDS-11136: If a DELETING container has a non-empty CLOSED replica, the container should be moved back to CLOSED
-      state.
+       * HDDS-11136: If a DELETING container has a non-empty CLOSED replica, the container should be moved back to
+       * CLOSED state.
+       *
+       * HDDS-12421: If a DELETED container has a non-empty CLOSED replica, the container should also be moved back to
+       * CLOSED state.
        */
+
+      boolean replicaIsEmpty = replica.hasIsEmpty() && replica.getIsEmpty();
+      // If container is in DELETED state and the reported replica is empty, delete the empty replica.
+      if (container.getState() == HddsProtos.LifeCycleState.DELETED && replicaIsEmpty) {
+        deleteReplica(containerId, datanode, publisher, "DELETED");
+        ignored = true;
+        break;
+      }
+
       boolean replicaStateAllowed = replica.getState() == State.CLOSED;
       boolean replicaNotEmpty = replica.hasIsEmpty() && !replica.getIsEmpty();
       if (replicaStateAllowed && replicaNotEmpty) {
-        logger.info("Moving DELETING container {} to CLOSED state, datanode {} reported replica with state={}, " +
-            "isEmpty={} and keyCount={}.", containerId, datanode.getHostName(), replica.getState(), false,
-            replica.getKeyCount());
-        containerManager.transitionDeletingToClosedState(containerId);
+        logger.info("Moving container {} from {} to CLOSED state, datanode {} reported replica with state={}, " +
```

Comment:
How about other state  (! replicaStateAllowed ) for DELETED state ? Do those need trigger move back like,
OPEN/CLOSING/QUASI_CLOSED/UNHEALTHY ....

Do same needs handling for DELETING state also? may be missing.

Reply:
I don't think containers in these states would currently be reported to SCM anyways, but I think `isEmpty` is more robust than any state check in this context so that is all we should use.